### ISSUE TITLE
Bug in Truncate Ternary ConditionalNode Optimisations

### ISSUE THREAD

**Issue by default-jamc** [oracle/graal#5101]
# Overview
In ConditionalNode there's an optimisation for the expressions: 

`x < 0.0 ? ceil(x) : floor(x)` 
`0.0 < x ? floor(x) : ceil(x)`

returning `RoundNode(((RoundNode) trueValue).value, RoundingMode.TRUNCATE)`

I noticed that in the conditions leading up to this transformation, there isn't a check for whether the `x` in `x < 0.0` and `0.0 < x` is the same `x` as within the `ceil(x)` and `floor(x)` calls, which could result in the wrong branch being run. 

## Example Program
I've attached an example program which illustrates the `y < 0.0` version of the bug:

```java
public class TruncateTernary {

  public static void main(String[] args) {
    double result = 0.0;
    double result2 = 0.0;
    float last_x = 0.0f;
    for (float x = 0.6f; x < 16_777_215.0f; x += 1.0f) {
      last_x = x;
      result = truncateTernary1(x, -x);
      assert result == Math.ceil(x) : "x=" + x + " result=" + result;
      result2 = truncateTernary1(x, x);
      assert result2 == Math.floor(x);
    }
    System.out.println("truncate with " + last_x + " gives " + result + " , " + result2);
  }

  public static double truncateTernary1(float x, float y) {
    return (y < 0.0 ? Math.ceil(x) : Math.floor(x));
  }
}
```

This runs correctly with a standard JDK and prints:

`truncate(1.6777214E7, -1.6777214E7) = 1.6777214E7`

but when compiled and run with GraalVM CE 22.2 it gives this exception:
```
% javac TruncateTernary.java 
% java -ea TruncateTernary   
Exception in thread "main" java.lang.AssertionError: x=421314.6 result=421314.0
      at TruncateTernary.main(TruncateTernary.java:10)
```

## Potential Fix
I've written a potential fix which includes a condition to check that the `x` in the condition is the same as the `x` in the `floor(x)` and `ceil(x)` calls. The diff for `ConditionalNode` is attached below:

```java
                         (
                                 // x < 0.0 ? ceil(x) : floor(x)
                                 ((FloatLessThanNode) condition).getY().isDefaultConstant() &&
+                                ((FloatLessThanNode) condition).getX().getClass() ==
+                                        (((RoundNode) trueValue).value).getClass() &&
+                                ((FloatLessThanNode) condition).getX().valueEquals(((RoundNode) trueValue).value) &&
                                 ((RoundNode) trueValue).mode() == RoundingMode.UP &&
                                 ((RoundNode) falseValue).mode() == RoundingMode.DOWN
                         ) || (
                                 // 0.0 < x ? floor(x) : ceil(x)
                                 ((FloatLessThanNode) condition).getX().isDefaultConstant() &&
+                                ((FloatLessThanNode) condition).getY().getClass() ==
+                                        (((RoundNode) trueValue).value).getClass() &&
+                                ((FloatLessThanNode) condition).getY().valueEquals(((RoundNode) trueValue).value) &&
                                 ((RoundNode) trueValue).mode() == RoundingMode.DOWN &&
                                 ((RoundNode) falseValue).mode() == RoundingMode.UP
                         )

```

## Unit Test
I've also written a unit test to check the behaviour of the code and whether the optimisation is applied: 

```java
package org.graalvm.compiler.core.test;

import org.graalvm.compiler.nodes.StructuredGraph;
import org.graalvm.compiler.nodes.calc.FloatLessThanNode;
import org.junit.Test;

public class ConditionalTruncateTernaryTest extends GraalCompilerTest {

    /** Tests for the TruncateTernary optimisations */
    public static double truncateTernary1(float x, float y) {
        return (y < 0.0 ? Math.ceil(x) : Math.floor(x));
    }

    public static double truncateTernary2(float x, float y) {
        return (0.0 < y ? Math.floor(x) : Math.ceil(x));
    }
    
    private void checkNodes(String methodName) {
        StructuredGraph graph = parseForCompile(getResolvedJavaMethod(methodName));
        createCanonicalizerPhase().apply(graph, getProviders());

        // If there are no FloatLessThanNodes in the graph, then the optimisation ran, and the test should fail.
        assertFalse(graph.getNodes().filter(FloatLessThanNode.class).count() == 0);

        // Testing optimised value against unoptimised value
        float y = 200.7f;
        float x = -2.3f;
        test(methodName, x, y);
    }

    @Test
    public void testConditional() {
        checkNodes("truncateTernary1");
        checkNodes("truncateTernary2");
    }
}

```

**Comment by gergo-** [comment]
Thank you for this report, @default-jamc . Your analysis of the problem is correct.
For the record, we can only accept contributions of code if you have signed the [Oracle Contributor Agreement (OCA)](https://oca.opensource.oracle.com/). In this particular case I will not use any of your code because the original code needs a bit more restructuring in my opinion.
Tracking this internally as GR-41636.

**Comment by gergo-** [comment]
Fixed by b07a99a.
